File: src-tauri/src/commands/external_access.rs
use serde::{Deserialize, Serialize};

use crate::{
    research::external_access::{
        ExternalAccessDestination, ExternalAccessError, ExternalAccessInput, open_in_system_browser,
    },
    system_browser::SystemBrowser,
};

/// Tagged external destination submitted for Rust validation and browser handoff.
#[derive(Debug, Deserialize, Eq, PartialEq)]
#[serde(tag = "destination", rename_all = "snake_case", deny_unknown_fields)]
pub(crate) enum OpenExternalAccessRequest {
    Publisher { url: String },
    Institutional { url: String },
    Doi { doi: String },
    GoogleScholar { query: String },
}

/// Stable acknowledgement returned after the OS accepts a browser handoff.
#[derive(Debug, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct OpenExternalAccessResponse {
    status: ExternalAccessStatus,
    destination: ExternalAccessDestination,
}

#[derive(Debug, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
enum ExternalAccessStatus {
    Opened,
}

/// Stable failures exposed by `open_external_access` over Tauri IPC.
#[derive(Debug, Eq, PartialEq, Serialize)]
#[serde(tag = "code", rename_all = "snake_case")]
pub(crate) enum OpenExternalAccessError {
    InvalidUrl,
    InvalidDoi,
    InvalidSearchQuery,
    BrowserUnavailable,
}

/// Opens one validated research destination in the default system browser.
#[tauri::command]
pub(crate) fn open_external_access(
    request: OpenExternalAccessRequest,
) -> Result<OpenExternalAccessResponse, OpenExternalAccessError> {
    let destination = open_in_system_browser(&SystemBrowser, request.into())?;
    Ok(OpenExternalAccessResponse::opened(destination))
}

impl OpenExternalAccessResponse {
    fn opened(destination: ExternalAccessDestination) -> Self {
        Self {
            status: ExternalAccessStatus::Opened,
            destination,
        }
    }
}

impl From<OpenExternalAccessRequest> for ExternalAccessInput {
    fn from(request: OpenExternalAccessRequest) -> Self {
        match request {
            OpenExternalAccessRequest::Publisher { url } => Self::PublisherUrl(url),
            OpenExternalAccessRequest::Institutional { url } => Self::InstitutionalUrl(url),
            OpenExternalAccessRequest::Doi { doi } => Self::Doi(doi),
            OpenExternalAccessRequest::GoogleScholar { query } => Self::GoogleScholarQuery(query),
        }
    }
}

impl From<ExternalAccessError> for OpenExternalAccessError {
    fn from(error: ExternalAccessError) -> Self {
        match error {
            ExternalAccessError::InvalidUrl => Self::InvalidUrl,
            ExternalAccessError::InvalidDoi => Self::InvalidDoi,
            ExternalAccessError::InvalidSearchQuery => Self::InvalidSearchQuery,
            ExternalAccessError::BrowserUnavailable => Self::BrowserUnavailable,
        }
    }
}

#[cfg(test)]
mod tests {
    use serde_json::json;

    use super::*;

    const TYPED_COMMAND: fn(
        OpenExternalAccessRequest,
    ) -> Result<OpenExternalAccessResponse, OpenExternalAccessError> = open_external_access;

    #[test]
    fn command_signature_is_typed() {
        let _ = TYPED_COMMAND;
    }

    #[test]
    fn request_deserialization_is_stable() {
        let requests = [
            json!({ "destination": "publisher", "url": "https://publisher.example" }),
            json!({ "destination": "institutional", "url": "https://library.example" }),
            json!({ "destination": "doi", "doi": "10.1000/example" }),
            json!({ "destination": "google_scholar", "query": "example title" }),
        ];
        for request in requests {
            assert!(serde_json::from_value::<OpenExternalAccessRequest>(request).is_ok());
        }

        for request in [
            json!({ "destination": "publisher", "url": "https://example.org", "extra": true }),
            json!({ "destination": "doi", "url": "https://doi.org/10.1000/example" }),
            json!({ "destination": "google_scholar" }),
            json!({ "destination": "embedded_browser", "url": "https://example.org" }),
        ] {
            assert!(serde_json::from_value::<OpenExternalAccessRequest>(request).is_err());
        }
    }

    #[test]
    fn response_serialization_is_stable() {
        let responses = [
            ExternalAccessDestination::Publisher,
            ExternalAccessDestination::Institutional,
            ExternalAccessDestination::Doi,
            ExternalAccessDestination::GoogleScholar,
        ]
        .map(OpenExternalAccessResponse::opened);

        assert_eq!(
            serde_json::to_value(responses).unwrap(),
            json!([
                { "status": "opened", "destination": "publisher" },
                { "status": "opened", "destination": "institutional" },
                { "status": "opened", "destination": "doi" },
                { "status": "opened", "destination": "google_scholar" }
            ])
        );
    }

    #[test]
    fn error_serialization_is_stable() {
        let errors = [
            OpenExternalAccessError::from(ExternalAccessError::InvalidUrl),
            OpenExternalAccessError::from(ExternalAccessError::InvalidDoi),
            OpenExternalAccessError::from(ExternalAccessError::InvalidSearchQuery),
            OpenExternalAccessError::from(ExternalAccessError::BrowserUnavailable),
        ];

        assert_eq!(
            serde_json::to_value(errors).unwrap(),
            json!([
                { "code": "invalid_url" },
                { "code": "invalid_doi" },
                { "code": "invalid_search_query" },
                { "code": "browser_unavailable" }
            ])
        );
    }
}

File: src/ipc/externalAccess.ts
import { invokeCommand } from "./client";
import { isRecord } from "./documentEnvelope";

export type ExternalAccessDestination =
  | "publisher"
  | "institutional"
  | "doi"
  | "google_scholar";

export type OpenExternalAccessRequest =
  | { destination: "publisher"; url: string }
  | { destination: "institutional"; url: string }
  | { destination: "doi"; doi: string }
  | { destination: "google_scholar"; query: string };

export interface OpenExternalAccessResponse {
  status: "opened";
  destination: ExternalAccessDestination;
}

export type ExternalAccessCommandErrorCode =
  | "invalid_url"
  | "invalid_doi"
  | "invalid_search_query"
  | "offline"
  | "connectivity_unavailable"
  | "browser_unavailable";

export type ExternalAccessClientError =
  | { type: "command"; code: ExternalAccessCommandErrorCode }
  | { type: "invalid-response" }
  | { type: "transport" };

export type ExternalAccessResult =
  | { status: "opened"; destination: ExternalAccessDestination }
  | { status: "error"; error: ExternalAccessClientError };

const COMMAND_NAME = "open_external_access";

/** Requests a validated Rust-owned handoff to the default system browser. */
export async function openExternalAccess(
  request: OpenExternalAccessRequest,
): Promise<ExternalAccessResult> {
  try {
    const response = await invokeCommand<unknown>(COMMAND_NAME, { request });
    return resultFromResponse(response, request.destination);
  } catch (error: unknown) {
    return { status: "error", error: clientErrorFrom(error) };
  }
}

function resultFromResponse(
  response: unknown,
  requestedDestination: ExternalAccessDestination,
): ExternalAccessResult {
  if (
    !isOpenExternalAccessResponse(response) ||
    response.destination !== requestedDestination
  ) {
    return { status: "error", error: { type: "invalid-response" } };
  }
  return { status: "opened", destination: response.destination };
}

function clientErrorFrom(error: unknown): ExternalAccessClientError {
  if (isRecord(error) && hasExactFields(error, ["code"]) && isCommandErrorCode(error.code)) {
    return { type: "command", code: error.code };
  }
  return { type: "transport" };
}

function isOpenExternalAccessResponse(value: unknown): value is OpenExternalAccessResponse {
  return (
    isRecord(value) &&
    hasExactFields(value, ["status", "destination"]) &&
    value.status === "opened" &&
    isExternalAccessDestination(value.destination)
  );
}

function isExternalAccessDestination(value: unknown): value is ExternalAccessDestination {
  return (
    value === "publisher" ||
    value === "institutional" ||
    value === "doi" ||
    value === "google_scholar"
  );
}

function isCommandErrorCode(value: unknown): value is ExternalAccessCommandErrorCode {
  return (
    value === "invalid_url" ||
    value === "invalid_doi" ||
    value === "invalid_search_query" ||
    value === "offline" ||
    value === "connectivity_unavailable" ||
    value === "browser_unavailable"
  );
}

function hasExactFields(value: Record<string, unknown>, fields: string[]) {
  const keys = Object.keys(value);
  return keys.length === fields.length && fields.every((field) => keys.includes(field));
}
